Data/ByteString/Lazy/ReadNat.hs

```diff
@@ -0,0 +1,209 @@
+    {-# INLINE unrefEmpty #-}
+    unrefEmpty n s = let rest = BU.unsafeDrop n s
+        in if B.null rest then empty else rest
+#else
+_readNatural kont !bs = loop bs False 1 0 []
+  where
+    -- All done
+    loop s@Empty !valid !cnt !acc ns
+        | valid = let !r = kont (combine cnt acc ns) in Just (r, s)
+        | otherwise = Nothing
+    -- skip empty segment
+    loop (Chunk (BI.BS _ 0) cs) !valid !cnt !acc ns
+        = loop cs valid cnt acc ns
+    loop s@(Chunk c@(BI.BS _ !len) cs) !valid !cnt !acc ns
+        = case natdigits c cnt acc ns of
+            Result 0 _ _ _ -- no more digits found
+                | valid
+                  -> let !r = kont (combine cnt acc ns) in Just (r, s)
+                | otherwise
+                  -> Nothing
+            Result used d a ms
+                | used < len -- input not entirely digits
+                  -> let !c' = BU.unsafeDrop used c
+                         s'  = Chunk c' cs
+                         !r = kont (combine d a ms)
+                      in Just (r, s')
+                | otherwise -- perhaps more digits from the remaining chunks
+                  -> loop cs True d a ms
+#endif
+
+----- Misc functions
+
+-- | Largest decimal digit count that never overflows the accumulator
+--
+safeLog :: Word
+#if WORD_SIZE_IN_BITS == 32
+safeLog = 9
+#elif WORD_SIZE_IN_BITS == 64
+safeLog = 19
+#endif
+
+-- | 10-power base for little-ending sequence of ~Word-sized "digits"
+-- For n*log(n) cost, these are recursively merged pairwise, while
+-- squaring the base.
+safeBase :: Natural
+safeBase = 10 ^ safeLog
+
+-- The intermediate representation is a little-endian sequence in base @10^19@,
+-- prefixed by an initial element in base @10^cnt@ for some `cnt` between 1 and
+-- 19.  The final result is obtained by recursive pairwise merging of the tail
+-- followed by a final multiplication by 10^cnt and addition of the head.
+--
+natdigits :: BI.ByteString -- ^ Input chunk
+          -> Word         -- ^ partial digit count
+          -> Word         -- ^ accumulated element
+          -> [Natural]    -- ^ accumulated MSB elements
+          -> Result
+{-# INLINE natdigits #-}
+natdigits (BI.BS !fp !len) = \ cnt acc ns ->
+    BI.accursedUnutterablePerformIO $
+        BI.unsafeWithForeignPtr fp $ \ ptr -> do
+            let end = ptr `plusPtr` len
+            go ptr end cnt acc ns ptr
+  where
+    go start end = loop
+      where
+        loop :: Word -> Word -> [Natural] -> Ptr Word8 -> IO Result
+        loop !cnt !acc ns !ptr
+            | ptr == end = return $ Result (ptr `minusPtr` start) cnt acc ns
+            | otherwise = fromDigit <$> peek ptr >>= \ !d ->
+                if d > 9
+                then return $ Result (ptr `minusPtr` start) cnt acc ns
+                else if cnt < safeLog
+                then loop (cnt+1) (10*acc + d) ns $ ptr `plusPtr` 1
+                else loop 1 d (toNatural acc : ns) $ ptr `plusPtr` 1
```

Comment:
I'll can it, but the accumulator has a `BangPattern` for that argument, so I don't think this is necessary.  It may even cause redundant code to be generated, though ideally GHC would avoid that.

Am I missing something here?

Reply:
The list constructor `(:)` is lazy, so it's not quite redundant. The strictness annotation is intended to express the (non-trivial) knowledge that every element put in this list will eventually be evaluated, and thereby potentially avoid allocating a suspension. But looking more closely, both versions generate identical STG, Cmm, and assembly, so I guess this really doesn't accomplish anything. I'm not sure yet if GHC has decided that `wordToNatural#` is too inexpensive to be worth deferring or if there's something else going on.

Reply:
Good point, I didn't check the context.  Fixed.